### PR TITLE
add support for Time Formatting/Parsing

### PR DESCRIPTION
**Pull request by marek-veber**
Hi, I expect, that it's OK now.

I'm prepared to push next two pull requests, if You will find it usefull:
1) compatible changes to run on ruby1.8 and 1.9.1
2) files need to build .deb package (works on wheezy with ruby1.8 and 1.9.1 and libicu48 and on jessie with ruby2.1 and libicu52)

### CODE REVIEW

**Comment by jarib** [comment]
Thank you! Please fix the indentation to be 2 spaces only. It's messed up several places in this patch.

**Comment by marek-veber** [comment]
Ok, indentation is fixed.

File: lib/ffi-icu/time_formatting.rb [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+          end
+          out_ptr.string
+        rescue BufferOverflowError
+          raise BufferOverflowError, "needed: #{needed_length}" if retried
+          out_ptr = out_ptr.resized_to needed_length
+          retried = true
+          retry
+        end
+      end
+      # time-zone formating
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please add a blank line after the end here.

File: lib/ffi-icu/time_formatting.rb [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        @f = make_formatter(time_style, date_style, locale, time_zone)
+        if tz_style
+          f0 = get_date_format(true)
+          f1 = update_tz_format(f0, tz_style)    
+          if f1 != f0
+            set_date_format(true, f1)
+          end
+        end
+      end
+      def parse(str)
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please add a blank line after the `end`.

File: lib/ffi-icu/time_formatting.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        begin
+          Lib.check_error do |error|
+            case dt
+            when Date
+              needed_length = Lib.udat_format(@f, Time.mktime( dt.year, dt.month, dt.day, 0, 0, 0, 0 ).to_f * 1000.0, out_ptr, needed_length, nil, error)
+            when Time
+              needed_length = Lib.udat_format(@f, dt.to_f * 1000.0, out_ptr, needed_length, nil, error)
+            end
+          end
+          out_ptr.string
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please add a blank line before the `end`, before the return value.

File: lib/ffi-icu/time_formatting.rb [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                                          #   the localized exemplar city name for the special zone Etc/Unknown is used as the fallback
+                                          #   (for example, "Unknown City"), such as Los Angeles
+            # see: http://unicode.org/reports/tr35/tr35-dates.html#Date_Format_Patterns
+          }
+          repl = $tz_map[tz_style]
+          raise 'no such tz_style' unless repl
+          tz.gsub!(/^(\s*)(.*?)(\s*)$/, '\1'+repl+'\3')
+        end
+        return pre + tz + suff
+        format
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Line 145 can't be reached and should be removed. 
The explicit `return` keyword is redundant and should be removed.

File: lib/ffi-icu/time_formatting.rb [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        end
+      end
+      # time-zone formating
+      def update_tz_format(format, tz_style)
+        return format if format !~ /(.*?)(\s*(?:[zZOVV]+\s*))(.*?)/
+        pre, tz, suff = $1, $2, $3
+        if tz_style == :none
+          tz = ((tz =~ /\s/) && !pre.empty? && !suff.empty?) ? ' ' : ''
+        else
+          $tz_map ||= {
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Don't use a global variable. If necessary, use a class instance variable for this.

File: spec/time_spec.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      t7 = Time.at(1427593043) # in TZ=Europe/Prague Time.mktime(2015, 03, 29, 03, 37, 23)
+      t8 = Time.at(1427596704) # in TZ=Europe/Prague Time.mktime(2015, 03, 29, 04, 38, 24)
+
+      f1 = TimeFormatting.create(:locale => 'cs_CZ', :zone => 'Europe/Prague', :date => :long , :time => :long, :tz_style => :localized_long)
+      it 'check date_format for lang=cs_CZ' do
+        f1.get_date_format(true).should eql "d. MMMM y H:mm:ss ZZZZ"
+        f1.get_date_format(false).should eql "d. MMMM y H:mm:ss ZZZZ"
+      end
+      it "for lang=cs_CZ zone=Europe/Prague" do
+        f1.should be_an_instance_of TimeFormatting::DateTimeFormatter
```
Comment (jarib) [inline spec/time_spec.rb]: Please add a blank line between `end` and `it.

File: spec/time_spec.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      f2 = TimeFormatting.create(:locale => 'en_US', :zone => 'Europe/Moscow', :date => :short , :time => :long, :tz_style => :generic_location)
+      cldr_version = Lib.cldr_version.to_s
+      en_tz  = "Moscow Time"
+      en_sep = ","
+      if cldr_version <= "2.0.1"
+        en_tz  = "Russia Time (Moscow)"
+        en_sep = ""
+      end
+      en_exp = "M/d/yy#{en_sep} h:mm:ss a VVVV"
+      it 'check date_format for lang=en_US' do
```
Comment (jarib) [inline spec/time_spec.rb]: Blank line before `it`

File: spec/time_spec.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        f2.get_date_format(true).should eql en_exp
+        f2.get_date_format(false).should eql en_exp
+      end
+      it "lang=en_US zone=Europe/Moscow" do
```
Comment (jarib) [inline spec/time_spec.rb]: Ditto

File: lib/ffi-icu/time_formatting.rb [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+          end
+          out_ptr.string
+        rescue BufferOverflowError
+          raise BufferOverflowError, "needed: #{needed_length}" if retried
+          out_ptr = out_ptr.resized_to needed_length
+          retried = true
+          retry
+        end
+      end
+      def set_date_format(localized, pattern_str)
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please add a blank line between `end` and `def`.

File: lib/ffi-icu/time_formatting.rb [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+          d_len = time_zone_str.size
+        else
+          Lib.check_error { | error| 
+            i_len  = 150
+            time_zone = UCharPointer.new(i_len)
+            d_len = Lib.ucal_getDefaultTimeZone(time_zone, i_len, error) 
+          }
+        end
+        ptr = Lib.check_error { | error| Lib.udat_open(time_style, date_style, locale, time_zone, d_len, FFI::MemoryPointer.new(4), -1, error) }
+        FFI::AutoPointer.new(ptr, Lib.method(:udat_close))
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please add a blank line before the return value.

File: lib/ffi-icu/time_formatting.rb (hunk 5 above)
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Seems like this never changes, so you can make it a constant inside this class.

File: lib/ffi-icu/time_formatting.rb [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+          }
+          repl = $tz_map[tz_style]
+          raise 'no such tz_style' unless repl
+          tz.gsub!(/^(\s*)(.*?)(\s*)$/, '\1'+repl+'\3')
+        end
+        return pre + tz + suff
+        format
+      end
+
+      def get_date_format(localized=true)
```
Comment (jarib) [inline lib/ffi-icu/time_formatting.rb]: Please remove the `get_` prefix - it's very rare / should be avoided in Ruby.

**Comment by jarib** [comment]
Thanks. I've read through and added some nits, mostly code style.

**Comment by marek-veber** [comment]
ok - done

**Comment by jarib** [comment]
Travis is failing. Looks like you forgot `get_date_format` in some tests.

**Comment by marek-veber** [comment]
done

**Comment by jarib** [comment]
Thanks again!

Locally, I get this failure. I'm in CEST.

```
  1) ICU::TimeFormatting the TimeFormatting  lang=en_US zone=Europe/Moscow
     Failure/Error: f2.format(t0).should eql "11/12/15#{en_sep} 6:21:16 PM #{en_tz}"

       expected: "11/12/15, 6:21:16 PM Moscow Time"
            got: "11/12/15, 5:21:16 PM Moscow Time"

       (compared using eql?)
     # /opt/rubies/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-expectations-2.14.5/lib/rspec/expectations/fail_with.rb:32:in `fail_with'
     # /opt/rubies/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-expectations-2.14.5/lib/rspec/expectations/handler.rb:34:in `handle_matcher'
     # /opt/rubies/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-expectations-2.14.5/lib/rspec/expectations/syntax.rb:53:in `should'
     # ./spec/time_spec.rb:54:in `block (3 levels) in <module:ICU>'
```

Can you look into it?

**Comment by marek-veber** [comment]
I'have the same primary time zone. Please can You send me Your local info:

irb(main):013:0> ICU::Lib.cldr_version.to_s
=> "2.0.1"

irb(main):014:0> ICU::Lib.version.to_s
=> "4.8.1.1"

irb(main):001:0> Time.at(1447338076)
=> 2015-11-12 15:21:16 +0100

 Thanks.

**Comment by jarib** [comment]
```irb
>> ICU::Lib.cldr_version.to_s
=> "25.0"
>>  ICU::Lib.version.to_s
=> "53.1"
>>  Time.at(1447338076)
=> 2015-11-12 15:21:16 +0100
>>
```

**Comment by jarib** [comment]
This is Ruby 2.2.2 on OS X 10.10.4.

**Comment by marek-veber** [comment]
It's result dependent on libicu version.
On debian wheezy (libicu48) or debian jessie (libicu52) it's OK.

But, when I try to install libicu55 from https://packages.debian.org/experimental/i386/libicu55/download :
irb(main):006:0> ICU::Lib.cldr_version.to_s
=> "27.0.1"
irb(main):007:0> ICU::Lib.version.to_s
=> "55.1"

than:
rb(main):003:0> f2 = ICU::TimeFormatting.create(:locale => 'en_US', :zone => 'Europe/Moscow', :date => :short , :time => :long, :tz_style => :generic_location)
=> #<ICU::TimeFormatting::DateTimeFormatter:0x00000002bba8d0 @f=#<FFI::AutoPointer address=0x00000002e0e860>>
irb(main):004:0> t0 = Time.at(1447338076)
=> 2015-11-12 15:21:16 +0100

irb(main):005:0> f2.format(t0)
=> "11/12/15, 5:21:16 PM Moscow Time"

For example on libicu55 it's:
irb(main):045:0> f2.format(Time.parse "2014-10-25 00:00")
=> "10/25/14, 2:00:00 AM Moscow Time"
irb(main):046:0> f2.format(Time.parse "2014-10-26 00:00")
=> "10/26/14, 1:00:00 AM Moscow Time"

On libicu52 i'ts:
irb(main):010:0> f2.format(Time.parse "2014-10-25 00:00")
=> "10/25/14, 2:00:00 AM Moscow Time"
irb(main):011:0> f2.format(Time.parse "2014-10-26 00:00")
=> "10/26/14, 2:00:00 AM Moscow Time"

So there is change in timezone definition for Moscow Time starting cca 2014-10-26 in version libicu53,
so we can change the tested values moving 2year back to works with new versions of libicu TZ definitions.

**Comment by jarib** [comment]
Sounds good to me. Can you open a new PR that does that?

**Pull request merged**